src/esp/sim/Simulator.cpp

```diff
@@ -833,5 +834,58 @@ void Simulator::setObjectLightSetup(int objectID,
   }
 }
 
+int Simulator::findNearestObjectUnderCrosshair(int refObjectID, 
+    Magnum::Vector3 point, Magnum::Vector3 refPoint, const Magnum::Vector2i& viewSize, 
+    float distance) {
+  float best_fraction = 99999.0;
+  int nearestObjId = ID_UNDEFINED;
+  auto& sceneGraph = sceneManager_->getSceneGraph(activeSceneID_);
+  gfx::RenderCamera* renderCamera_ = &sceneGraph.getDefaultRenderCamera();
+
+  // try a ray test
+  physics::BulletPhysicsManager* bpm =
+      static_cast<physics::BulletPhysicsManager*>(physicsManager_.get());
+  Magnum::Vector3 cast =
+      (point - renderCamera_->node().absoluteTranslation()).normalized();
+  btCollisionWorld::AllHitsRayResultCallback hit =
+      bpm->castRay(renderCamera_->node().absoluteTranslation(), cast);
+  Magnum::Vector3 hitPoint;
+
+  for (int hitIx = 0; hitIx < hit.m_hitPointWorld.size(); hitIx++) {
+    // Corrade::Utility::Debug()
+    //     << " hit fraction: " << hit.m_hitFractions.at(hitIx);
+    if (hit.m_hitFractions.at(hitIx) < best_fraction) {
+      best_fraction = hit.m_hitFractions.at(hitIx);
+      
+      nearestObjId = bpm->getObjectIDFromCollisionObject(
+          hit.m_collisionObjects.at(hitIx));
+      hitPoint = Magnum::Vector3{hit.m_hitPointWorld.at(hitIx)};
+    }
+  }
+
+  if ((hitPoint - refPoint).length() > distance) {
+    return -1;
+  }
+
+  return nearestObjId;
+}
+
+Magnum::Vector3 Simulator::unproject(const Magnum::Vector2i& crosshairPos,
```

Comment:
See #699 for where these functions should live: https://github.com/facebookresearch/habitat-sim/pull/699